homeassistant/components/sleepiq/binary_sensor.py

```diff
@@ -34,16 +48,20 @@ class IsInBedBinarySensor(SleepIQSensor, BinarySensorEntity):
 
     def __init__(
         self,
-        coordinator: SleepIQDataUpdateCoordinator,
-        bed_id: str,
-        side: str,
+        coordinator: DataUpdateCoordinator,
+        bed: SleepIQBed,
+        sleeper: SleepIQSleeper,
     ) -> None:
-        """Initialize the SleepIQ bed side binary sensor."""
-        super().__init__(coordinator, bed_id, side, IS_IN_BED)
-
-    @callback
-    def _async_update_attrs(self) -> None:
-        """Update sensor attributes."""
-        super()._async_update_attrs()
-        self._attr_is_on = getattr(self.side_data, IS_IN_BED)
-        self._attr_icon = ICON_OCCUPIED if self.is_on else ICON_EMPTY
```

Comment:
Please keep the existing `_async_update_attrs` methods

You are setting `_attr_icon` and the `icon` property here

Reply:
The problem I saw with the _async_update_attrs  function is it was only being updated when the coordinator would update which meant that there was a brief period where the icon would be not accurate when the platform was created after the coordinator did the first update.  The underlying data  of the icon and state properties would be updated, but since the attributes are only set on update, the value in HA would be out of date

Reply:
Was `_async_update_attrs` not being called in `__init__` to set the `_attrs` so they are set correctly when the entity is added to hass ?

Reply:
It wasn't. I thought about adding it, but figured having two extra callback functions plus extraneous function calls was more overhead than was needed to just check a boolean and it looked like only 2 other integrations were using this update_attrs style so that didn't seem to be the standard

Reply:
Updating _attrs is newer and preferred since it ends up being a smaller when we have additional platforms. Binary sensor and sensor probably aren't the best example of this however since they are one of the more simple platforms.  It really starts to make a difference with light, media player, and climate platforms

Reply:
For context we just updated WiZ to use _attrs when updating state

Before
https://github.com/home-assistant/core/blob/432d9a8f19454628d815152516cac0ee24f81420/homeassistant/components/wiz/light.py#L75

After
https://github.com/home-assistant/core/blob/445ad1d592588fd91442ef45fe196ec3272dec44/homeassistant/components/wiz/light.py#L100

Which made adding switch support later really really simple
https://github.com/home-assistant/core/blob/dev/homeassistant/components/wiz/switch.py#L30

Reply:
Why does it override `_handle_coordinator_update` instead of just adding `_async_update_attrs` as another listener for the Coordinator?

Reply:
Its using `CoordinatorEntity` as a base class https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/update_coordinator.py#L323

Since `self.coordinator.async_add_listener` doesn't guarantee that listeners will fire in a specific order you might end up writing state https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/update_coordinator.py#L325 before the attrs are updated.

Reply:
I see.  Looks like `CoordinatorEntity` should be enhanced to have a standard function that would be called when the data is updated that calls `async_write_ha_state`, but that's out of the scope of this PR